Sources/StreamChat/Controllers/ChannelListController/ChannelListController.swift

```diff
@@ -164,18 +185,32 @@ public class ChatChannelListController: DataController, DelegateCallable, DataSt
             log.error("Failed to perform fetch request with error: \(error). This is an internal error.")
         }
     }
+
+    private func channelBelongsToController(_ channel: ChatChannel) -> Bool {
+        if let filterBlock = filterBlock {
+            return filterBlock(channel)
+        }
+        do {
+            return try channel.meets(query.filter)
+        } catch {
+            log.error("Unable to resolve complex filter. Please pass a `filterBlock` when intializing ChatChannelListController")
+        }
+        return true
+    }
     
     private func handleUnlinkedChannels(_ changes: [ListChange<ChatChannel>]) {
         guard state == .remoteDataFetched else {
             log.debug("Ignoring inserted/updated unlinked channels due to query \(query) not being synced.")
             return
         }
-        
+
         let channels = changes.compactMap { change -> ChatChannel? in
             switch change {
             case let .insert(channel, _):
+                guard channelBelongsToController(channel) else { return nil }
                 return (delegate?.controller(self, shouldAddNewChannelToList: channel) ?? true) ? channel : nil
             case let .update(channel, _):
+                guard channelBelongsToController(channel) else { return nil }
                 return (delegate?.controller(self, shouldListUpdatedChannel: channel) ?? true) ? channel : nil
```

Comment:
Does this break the current behaviour somehow? We need to make sure the previous behaviour is still the default, otherwise, we might break some customers 🤔 

This query problem is getting a bit out of control. From what I understood, the problem is that we can't currently link the query to a controller right? 🤔 (I'm not sure even, I'm confused)

Possible solutions:
- Since we can't make the queries equatable because of their complexity, would making the queries reference types instead of value types, help here? 🤔 
- Another solution could be to have the query contain the object address of the ChannelController so that we know what channel controller does the query belong to.

Reply:
It does not break the current behaviour, not. That is because the default implementation of `channelBelongsToController` just returns true, unless a filterBlock is passed.

The problem around query is not that we cannot link a query to a controller, but the fact that we are using query's linked channels as a source of data for our channel list. So the solutions posted below do not solve the issue.

I will shoot a Notion doc to make sure the issues are understood 🙏

Reply:
Thanks for the explanation 🙏